File: 백준/Gold/10986. 나머지 합/나머지 합.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        
        StringTokenizer st = new StringTokenizer(br.readLine());
        int n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        long[] arr = new long[n+1];
        long[] div = new long[m];
        long cnt = 0;
        st = new StringTokenizer(br.readLine());
        for (int i = 1; i <= n; i++) {
            long val = Long.parseLong(st.nextToken());
            arr[i] = arr[i-1] + val;
            int remainder = (int) (arr[i] % m);
            if (remainder == 0) cnt++;
            div[remainder] = div[remainder] + 1;
        }
        
        for (int i = 0; i < m; i++) {
            if (div[i] > 1) {
                cnt += div[i] * (div[i]-1) / 2;
            }
        }
        
        
        bw.write(cnt + "\n");
        bw.flush();
        bw.close();
    }
}

File: 백준/Gold/10986. 나머지 합/나머지 합.py
import sys

input = sys.stdin.readline
print = sys.stdout.write

n, m = map(int, input().split())

arr = [0] * (n+1)
div = [0] * m
cnt = 0
values = list(map(int, input().split()))
for i in range(1, n+1):
    val = values[i-1]
    arr[i] = arr[i-1] + val
    remainder = arr[i] % m
    if (remainder == 0):
        cnt = cnt + 1
    div[remainder]= div[remainder] + 1

for i in range (m):
    if (div[i] > 1):
        cnt += div[i] * (div[i] -1) // 2

    
print(str(cnt))